src/main/java/io/opentracing/contrib/global/GlobalTracer.java

```diff
@@ -0,0 +1,184 @@
+package io.opentracing.contrib.global;
+
+import io.opentracing.NoopTracerFactory;
+import io.opentracing.Span;
+import io.opentracing.SpanContext;
+import io.opentracing.Tracer;
+import io.opentracing.contrib.global.concurrent.TracedCallable;
+import io.opentracing.contrib.global.concurrent.TracedRunnable;
+import io.opentracing.propagation.Format;
+
+import java.util.Iterator;
+import java.util.ServiceLoader;
+import java.util.concurrent.Callable;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * The {@linkplain GlobalTracer} forwards all methods to a single {@link Tracer} implementation that can be
+ * instantiated in one of two ways:
+ * <ol>
+ * <li>Explicitly, by calling {@link #setTracer(Tracer)} with a configured tracer implementation, or:</li>
+ * <li>Automatically by using the Java <code>ServiceLoader</code> SPI mechanism to load an implementation
+ * from the classpath.</li>
+ * </ol>
+ *
+ * @author Sjoerd Talsma
+ */
+public final class GlobalTracer implements Tracer {
+    private static final Logger LOGGER = Logger.getLogger(GlobalTracer.class.getName());
+
+    private static final GlobalTracer INSTANCE = new GlobalTracer();
+
+    /**
+     * The resolved {@link Tracer} to delegate the global tracing implementation to.<br>
+     * This can be either an {@link #setTracer(Tracer) explicitly set delegate}
+     * or the automatically resolved Tracer implementation.<br>
+     * Management of this reference is the responsibility of the {@link #lazyTracer()} method.
+     */
+    private final AtomicReference<Tracer> globalTracer = new AtomicReference<Tracer>();
```

Comment:
you can test this by making the ref package protected. set it to null and control in the tests.

To test concurrency, you can extract loadSingleton into a static package-private type, which is assigned in a package-protected constructor. In your concurrency tests, you can change that to something similar to a number generator and assert that even when 10 threads simultaneously call this, the assignment happens once.

The default LoadSingletonFromServiceLoader implementation (the package-private class) would expose a method that calls `ServiceLoader.load(Tracer.class, Tracer.class.getClassLoader())`. In your tests, you'd override that method so that serviceloader isn't needed to test all the edge cases your code is attempting to solve.

The above reduces the untested code mainly to the serviceloader hook.

Freedom and Responsibility.. feel free to make fancy code, but.. gotta test it!